heat/core/manipulations.py

```diff
@@ -4178,3 +4179,97 @@ def mpi_topk(a, b, mpi_type):
+    if size > a.shape[dimension]:  # size too large, runtime error or value error?
+        raise ValueError(
+            f"maximum size for DNDarray at dimension {dimension} is {a.shape[dimension]} but size is {size}."
+        )
+
+    comm = a.comm
+    dev = a.device
+    tdev = dev.torch_device
+
+    if a.split is None or comm.size == 1 or a.split != dimension:  # early out
+        ret = factories.array(
+            a.larray.unfold(dimension, size, step), is_split=a.split, device=dev, comm=comm
+        )
+
+        return ret
+    else:  # comm.size > 1 and split axis == unfold axis
+        # index range [0:sizedim-1-(size-1)] = [0:sizedim-size]
+        # --> size of axis: ceil((sizedim-size+1) / step) = floor(sizedim-size) / step)) + 1
+        # ret_shape = (*a_shape[:dimension], int((a_shape[dimension]-size)/step) + 1, a_shape[dimension+1:], size)
+
+        if (size - 1 > a.lshape_map[:, dimension]).any():
+            raise RuntimeError("Chunk-size needs to be at least size - 1.")
+        a.get_halo(size - 1)
+        a_lshapes_cum = torch.hstack(
+            [
+                torch.zeros(1, dtype=torch.int32, device=tdev),
+                torch.cumsum(a.lshape_map[:, dimension], 0),
+            ]
+        )
+        # min local index in unfold dimension
+        min_index = ((a_lshapes_cum[comm.rank] - 1) // step + 1) * step - a_lshapes_cum[comm.rank]
+        loc_unfold_shape = list(a.lshape)
+        loc_unfold_shape[dimension] -= min_index
+        if loc_unfold_shape[dimension] >= size:  # some unfold arrays are unfolds of the local array
+            unfold_loc = a.larray[
+                dimension * (slice(None, None, None),) + (slice(min_index, None, None), Ellipsis)
+            ].unfold(dimension, size, step)
+        else:
+            loc_unfold_shape[dimension] = 0
+            unfold_loc = torch.zeros((*loc_unfold_shape, size), device=tdev)
+        ret_larray = unfold_loc
+        if (
+            comm.rank < comm.size - 1 and min_index < a.lshape[dimension]
+        ):  # possibly unfold with halo from next rank
+            max_index = a.lshape[dimension] - min_index - 1
+            max_index = max_index // step * step + min_index  # max local index in unfold dimension
+            rem = max_index + size - a.lshape[dimension]
+            if rem > 0:  # need data from halo
+                unfold_halo = torch.cat(
+                    (
+                        a.larray[
+                            dimension * (slice(None, None, None),)
+                            + (slice(max_index, None, None), Ellipsis)
+                        ],
+                        a.halo_next[
+                            dimension * (slice(None, None, None),)
+                            + (slice(None, rem, None), Ellipsis)
+                        ],
+                    ),
+                    dimension,
+                ).unfold(dimension, size, step)
```

Comment:
I think this might not really make things easier because as far as I understand it I would then have to make a case distinction to remove the ´halo.prev´ for the nodes in the middle. Is there any way to only append the ´halo.next´?

Reply:
does #1509 help?

Reply:
That should be exactly what I needed, I'll merge it into this PR to test it out.